geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java

```diff
@@ -0,0 +1,525 @@
+    vms.forEach(vm -> vm.invoke(() -> {
+      waitForSilence();
+      Region<String, String> region = cacheRule.getCache().getRegion(REGION_NAME);
+
+      assertThat(region.isEmpty()).isTrue();
+    }));
+  }
+
+  /**
+   * Waits until there are no partition messages in flight and gets a snapshot of the region.
+   */
+  private Map<String, String> waitForSilenceAndGetRegionSnapshot() {
+    waitForSilence();
+    Map<String, String> regionSnapshot = new HashMap<>();
+    Region<String, String> region = cacheRule.getCache().getRegion(REGION_NAME);
+    region.forEach(regionSnapshot::put);
+
+    return regionSnapshot;
+  }
+
+  /**
+   * Asserts that the region data is consistent on requested VMs.
+   */
+  private void assertRegionDataConsistency(List<VM> vms) {
+    // Get first snapshot to compare with the rest.
+    final Map<String, String> vm0Snapshot =
+        vms.get(0).invoke(this::waitForSilenceAndGetRegionSnapshot);
+
+    vms.forEach(vm -> vm.invoke(() -> {
+      final Map<String, String> thisSnapshot = waitForSilenceAndGetRegionSnapshot();
+      assertThat(thisSnapshot).isEqualTo(vm0Snapshot);
+      thisSnapshot.forEach((key, value) -> assertThat(value).isEqualTo("Value_" + key));
+    }));
+  }
+
+  /**
+   * Populates the region and verifies the data on the selected VMs.
+   */
+  private void populateRegion(VM feeder, int entryCount, List<VM> vms) {
+    feeder.invoke(() -> {
+      Region<String, String> region = cacheRule.getCache().getRegion(REGION_NAME);
+      IntStream.range(0, entryCount).forEach(i -> region.put(String.valueOf(i), "Value_" + i));
+    });
+
+    assertRegionData(entryCount, vms);
+  }
+
+  /**
+   * Continuously execute get operations on the PartitionedRegion for the given durationInSeconds.
+   */
+  private void executeGets(final int numEntries, final long durationInSeconds) {
+    Cache cache = cacheRule.getCache();
+    Region<String, String> region = cache.getRegion(REGION_NAME);
+    Instant finishTime = Instant.now().plusSeconds(durationInSeconds);
+
+    while (Instant.now().isBefore(finishTime)) {
+      // Region might have been cleared in between, that's why we check for null.
+      IntStream.range(0, numEntries).forEach(i -> {
+        Optional<String> nullableValue = Optional.ofNullable(region.get(String.valueOf(i)));
+        nullableValue.ifPresent(value -> assertThat(value).isEqualTo("Value_" + i));
+      });
+    }
+  }
+
+  /**
+   * Continuously execute put operations on the PartitionedRegion for the given durationInSeconds.
+   */
+  private void executePuts(final int numEntries, final long durationInSeconds) {
+    Cache cache = cacheRule.getCache();
+    Region<String, String> region = cache.getRegion(REGION_NAME);
+    Instant finishTime = Instant.now().plusSeconds(durationInSeconds);
+
+    while (Instant.now().isBefore(finishTime)) {
+      IntStream.range(0, numEntries).forEach(i -> region.put(String.valueOf(i), "Value_" + i));
```

Comment:
it's better to change the continuous put into continuous putAll.

Reply:
Any reason for that?. I can certainly do it, just wondering what's the actual reasoning behind the request.

Reply:
putAll will have longer window to reproduce synchronization issues

Reply:
Cool, will add another test to execute multiple `putAll` instead of regular `put` opertations.